{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pybiomart.dataset import Dataset\n",
    "from tfac.dataOrg import extractData\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = extractData('tfac/data/geneIdentifiers.xlsx', 'A:D')\n",
    "data = data.to_numpy()\n",
    "ranges = np.array([21337, 56202, 23316])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ZZZ3'"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data[ranges[0]-1][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "methylation = np.append(data[:12158,0:2],data[12159:21338,0:2], axis = 0).astype(str)\n",
    "geneExp = data[:,2].astype(str)\n",
    "copyNum = data[:23316,3].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENSG00000186205\n"
     ]
    }
   ],
   "source": [
    "print(geneExp[0].split('.')[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "geneEntrez = np.zeros(ranges[1]).astype(str)\n",
    "for i in range(ranges[1]):\n",
    "    entrez = dataset.query(attributes=['entrezgene_id'],\n",
    "              filters={'link_ensembl_gene_id': [geneExp[i].split('.')[0]]})\n",
    "    if entrez.empty:\n",
    "        geneEntrez[i] = 'nan'\n",
    "    else:\n",
    "        geneEntrez[i] = entrez.values[0][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8104\n"
     ]
    }
   ],
   "source": [
    "for count, i in enumerate(geneEntrez):\n",
    "    if i == '0.0':\n",
    "        print(count)\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['64757' '55016' '54996' ... '0.0' '0.0' '0.0']\n"
     ]
    }
   ],
   "source": [
    "print(geneEntrez)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENSG00000186205\n"
     ]
    }
   ],
   "source": [
    "print(data[0][2].split('.')[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset = Dataset(name='hsapiens_gene_ensembl', host='http://www.ensembl.org')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(0, 1)\n",
      "fuck\n"
     ]
    }
   ],
   "source": [
    "ch4 = dataset.query(attributes=['entrezgene_id'],\n",
    "              filters={'link_ensembl_gene_id': ['ENSG00000173077']})\n",
    "print(ch4.shape)\n",
    "if ch4.empty:\n",
    "    print('fuck')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1, 1)\n"
     ]
    }
   ],
   "source": [
    "ch4 = dataset.query(attributes=['entrezgene_id'],\n",
    "              filters={'link_ensembl_gene_id': ['ENSG00000100167']})\n",
    "print(ch4.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>type</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>link_so_mini_closure</td>\n",
       "      <td>list</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>link_go_closure</td>\n",
       "      <td>text</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>link_ensembl_transcript_stable_id</td>\n",
       "      <td>text</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>text</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>transcript_id</td>\n",
       "      <td>text</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>link_ensembl_gene_id</td>\n",
       "      <td>text</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chromosome_name</td>\n",
       "      <td>text</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>start</td>\n",
       "      <td>text</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>end</td>\n",
       "      <td>text</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>band_start</td>\n",
       "      <td>drop_down_basic_filter</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>band_end</td>\n",
       "      <td>drop_down_basic_filter</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>marker_start</td>\n",
       "      <td>drop_down_basic_filter</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>marker_end</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>hsapiens_encode.type</td>\n",
       "      <td>text</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>hsapiens_encode.encode_region</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>strand</td>\n",
       "      <td>text</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chromosomal_region</td>\n",
       "      <td>text</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>id_list_xrefs_filters</td>\n",
       "      <td>boolean_list</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>id_list_limit_xrefs_filters</td>\n",
       "      <td>id_list</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>id_list_microarray_filters</td>\n",
       "      <td>boolean_list</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>id_list_limit_microarray_filters</td>\n",
       "      <td>id_list</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>transcript_count_greater_than</td>\n",
       "      <td>text</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>transcript_count_less_than</td>\n",
       "      <td>text</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>biotype</td>\n",
       "      <td>list</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>transcript_biotype</td>\n",
       "      <td>list</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>source</td>\n",
       "      <td>list</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>transcript_source</td>\n",
       "      <td>list</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>transcript_tsl</td>\n",
       "      <td>boolean</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>transcript_gencode_basic</td>\n",
       "      <td>boolean</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>transcript_appris</td>\n",
       "      <td>boolean</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>mane_select</td>\n",
       "      <td>boolean</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>phenotype_description</td>\n",
       "      <td>list</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>phenotype_source</td>\n",
       "      <td>list</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>go_clos</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>go_name</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>go_evidence_code</td>\n",
       "      <td>list</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>homolog_filters</td>\n",
       "      <td>boolean_list</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>id_list_protein_domain_and_feature_filters</td>\n",
       "      <td>boolean_list</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>id_list_limit_protein_domain_filters</td>\n",
       "      <td>id_list</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>germ_line_variation_source</td>\n",
       "      <td>list</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>somatic_variation_source</td>\n",
       "      <td>list</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>with_validated_snp</td>\n",
       "      <td>boolean</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>so_consequence_name</td>\n",
       "      <td>list</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          name                    type  \\\n",
       "0                         link_so_mini_closure                    list   \n",
       "1                              link_go_closure                    text   \n",
       "2            link_ensembl_transcript_stable_id                    text   \n",
       "3                                      gene_id                    text   \n",
       "4                                transcript_id                    text   \n",
       "5                         link_ensembl_gene_id                    text   \n",
       "6                              chromosome_name                    text   \n",
       "7                                        start                    text   \n",
       "8                                          end                    text   \n",
       "9                                   band_start  drop_down_basic_filter   \n",
       "10                                    band_end  drop_down_basic_filter   \n",
       "11                                marker_start  drop_down_basic_filter   \n",
       "12                                  marker_end                           \n",
       "13                        hsapiens_encode.type                    text   \n",
       "14               hsapiens_encode.encode_region                           \n",
       "15                                      strand                    text   \n",
       "16                          chromosomal_region                    text   \n",
       "17                       id_list_xrefs_filters            boolean_list   \n",
       "18                 id_list_limit_xrefs_filters                 id_list   \n",
       "19                  id_list_microarray_filters            boolean_list   \n",
       "20            id_list_limit_microarray_filters                 id_list   \n",
       "21               transcript_count_greater_than                    text   \n",
       "22                  transcript_count_less_than                    text   \n",
       "23                                     biotype                    list   \n",
       "24                          transcript_biotype                    list   \n",
       "25                                      source                    list   \n",
       "26                           transcript_source                    list   \n",
       "27                              transcript_tsl                 boolean   \n",
       "28                    transcript_gencode_basic                 boolean   \n",
       "29                           transcript_appris                 boolean   \n",
       "30                                 mane_select                 boolean   \n",
       "31                       phenotype_description                    list   \n",
       "32                            phenotype_source                    list   \n",
       "33                                     go_clos                           \n",
       "34                                     go_name                           \n",
       "35                            go_evidence_code                    list   \n",
       "36                             homolog_filters            boolean_list   \n",
       "37  id_list_protein_domain_and_feature_filters            boolean_list   \n",
       "38        id_list_limit_protein_domain_filters                 id_list   \n",
       "39                  germ_line_variation_source                    list   \n",
       "40                    somatic_variation_source                    list   \n",
       "41                          with_validated_snp                 boolean   \n",
       "42                         so_consequence_name                    list   \n",
       "\n",
       "   description  \n",
       "0               \n",
       "1               \n",
       "2               \n",
       "3               \n",
       "4               \n",
       "5               \n",
       "6               \n",
       "7               \n",
       "8               \n",
       "9               \n",
       "10              \n",
       "11              \n",
       "12              \n",
       "13              \n",
       "14              \n",
       "15              \n",
       "16              \n",
       "17              \n",
       "18              \n",
       "19              \n",
       "20              \n",
       "21              \n",
       "22              \n",
       "23              \n",
       "24              \n",
       "25              \n",
       "26              \n",
       "27              \n",
       "28              \n",
       "29              \n",
       "30              \n",
       "31              \n",
       "32              \n",
       "33              \n",
       "34              \n",
       "35              \n",
       "36              \n",
       "37              \n",
       "38              \n",
       "39              \n",
       "40              \n",
       "41              \n",
       "42              "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset.list_filters()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "attributes = dataset.attributes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "ch4 = dataset.query(attributes=['hgnc_symbol', 'start_position', 'end_position'],\n",
    "              filters={'link_ensembl_gene_id': ['ENSG00000186205'], })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  HGNC symbol  Gene start (bp)  Gene end (bp)\n",
      "0       MARC1        220786913      220819659\n"
     ]
    }
   ],
   "source": [
    "print(ch4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "start_position\n",
      "transcript_start\n",
      "transcription_start_site\n",
      "cdd_start\n",
      "gene3d_start\n",
      "hamap_start\n",
      "hmmpanther_start\n",
      "pfam_start\n",
      "pirsf_start\n",
      "prints_start\n",
      "scanprosite_start\n",
      "pfscan_start\n",
      "sfld_start\n",
      "smart_start\n",
      "superfamily_start\n",
      "tigrfam_start\n",
      "interpro_start\n",
      "mobidblite_start\n",
      "ncoils_start\n",
      "seg_start\n",
      "sifts_import_start\n",
      "signalp_start\n",
      "tmhmm_start\n",
      "structure_gene_chrom_start\n",
      "structure_transcript_chrom_start\n",
      "structure_transcription_start_site\n",
      "structure_5_utr_start\n",
      "structure_3_utr_start\n",
      "exon_chrom_start\n",
      "cdna_coding_start\n",
      "genomic_coding_start\n",
      "structure_cds_start\n",
      "homologs_start_position\n",
      "cabingdonii_homolog_chrom_start\n",
      "gagassizii_homolog_chrom_start\n",
      "mspretus_homolog_chrom_start\n",
      "vpacos_homolog_chrom_start\n",
      "mmmarmota_homolog_chrom_start\n",
      "pformosa_homolog_chrom_start\n",
      "ccanadensis_homolog_chrom_start\n",
      "bbbison_homolog_chrom_start\n",
      "uamericanus_homolog_chrom_start\n",
      "nvison_homolog_chrom_start\n",
      "capalliatus_homolog_chrom_start\n",
      "acarolinensis_homolog_chrom_start\n",
      "uparryii_homolog_chrom_start\n",
      "smerianae_homolog_chrom_start\n",
      "dnovemcinctus_homolog_chrom_start\n",
      "sformosus_homolog_chrom_start\n",
      "charengus_homolog_chrom_start\n",
      "cporosus_homolog_chrom_start\n",
      "lbergylta_homolog_chrom_start\n",
      "lcalcarifer_homolog_chrom_start\n",
      "lsdomestica_homolog_chrom_start\n",
      "spartitus_homolog_chrom_start\n",
      "rbieti_homolog_chrom_start\n",
      "ccaeruleus_homolog_chrom_start\n",
      "lcoronata_homolog_chrom_start\n",
      "gwilldenowi_homolog_chrom_start\n",
      "sbboliviensis_homolog_chrom_start\n",
      "ppaniscus_homolog_chrom_start\n",
      "caperea_homolog_chrom_start\n",
      "mundulatus_homolog_chrom_start\n",
      "hburtoni_homolog_chrom_start\n",
      "ogarnettii_homolog_chrom_start\n",
      "cintestinalis_homolog_chrom_start\n",
      "csavignyi_homolog_chrom_start\n",
      "celegans_homolog_chrom_start\n",
      "ccapucinus_homolog_chrom_start\n",
      "fcatus_homolog_chrom_start\n",
      "pvitticeps_homolog_chrom_start\n",
      "cgobio_homolog_chrom_start\n",
      "ipunctatus_homolog_chrom_start\n",
      "ggallus_homolog_chrom_start\n",
      "nperdicaria_homolog_chrom_start\n",
      "ptroglodytes_homolog_chrom_start\n",
      "cgchok1gshd_homolog_chrom_start\n",
      "cgcrigri_homolog_chrom_start\n",
      "cgpicr_homolog_chrom_start\n",
      "psinensis_homolog_chrom_start\n",
      "atestudineus_homolog_chrom_start\n",
      "aocellaris_homolog_chrom_start\n",
      "gmorhua_homolog_chrom_start\n",
      "lchalumnae_homolog_chrom_start\n",
      "scanaria_homolog_chrom_start\n",
      "vursinus_homolog_chrom_start\n",
      "pcoquereli_homolog_chrom_start\n",
      "btaurus_homolog_chrom_start\n",
      "mfascicularis_homolog_chrom_start\n",
      "fdamarensis_homolog_chrom_start\n",
      "jhyemalis_homolog_chrom_start\n",
      "sdauricus_homolog_chrom_start\n",
      "odegus_homolog_chrom_start\n",
      "dclupeoides_homolog_chrom_start\n",
      "cldingo_homolog_chrom_start\n",
      "cfamiliaris_homolog_chrom_start\n",
      "ttruncatus_homolog_chrom_start\n",
      "eaasinus_homolog_chrom_start\n",
      "mleucophaeus_homolog_chrom_start\n",
      "dmelanogaster_homolog_chrom_start\n",
      "applatyrhynchos_homolog_chrom_start\n",
      "acalliptera_homolog_chrom_start\n",
      "eelectricus_homolog_chrom_start\n",
      "lafricana_homolog_chrom_start\n",
      "cmilii_homolog_chrom_start\n",
      "dnovaehollandiae_homolog_chrom_start\n",
      "mpfuro_homolog_chrom_start\n",
      "falbicollis_homolog_chrom_start\n",
      "trubripes_homolog_chrom_start\n",
      "tgelada_homolog_chrom_start\n",
      "nleucogenys_homolog_chrom_start\n",
      "chircus_homolog_chrom_start\n",
      "mauratus_homolog_chrom_start\n",
      "rroxellana_homolog_chrom_start\n",
      "mvitellinus_homolog_chrom_start\n",
      "ggorilla_homolog_chrom_start\n",
      "pmajor_homolog_chrom_start\n",
      "ahaastii_homolog_chrom_start\n",
      "sdumerili_homolog_chrom_start\n",
      "psimus_homolog_chrom_start\n",
      "cporcellus_homolog_chrom_start\n",
      "preticulata_homolog_chrom_start\n",
      "eburgeri_homolog_chrom_start\n",
      "eeuropaeus_homolog_chrom_start\n",
      "nmeleagris_homolog_chrom_start\n",
      "ecaballus_homolog_chrom_start\n",
      "hhucho_homolog_chrom_start\n",
      "bihybrid_homolog_chrom_start\n",
      "bthybrid_homolog_chrom_start\n",
      "pcapensis_homolog_chrom_start\n",
      "pranga_homolog_chrom_start\n",
      "omelastigma_homolog_chrom_start\n",
      "olhni_homolog_chrom_start\n",
      "olhsok_homolog_chrom_start\n",
      "olatipes_homolog_chrom_start\n",
      "cjaponica_homolog_chrom_start\n",
      "dordii_homolog_chrom_start\n",
      "pcinereus_homolog_chrom_start\n",
      "pmarinus_homolog_chrom_start\n",
      "lcrocea_homolog_chrom_start\n",
      "ppardus_homolog_chrom_start\n",
      "jjaculus_homolog_chrom_start\n",
      "etelfairi_homolog_chrom_start\n",
      "aowenii_homolog_chrom_start\n",
      "clanigera_homolog_chrom_start\n",
      "nbrichardi_homolog_chrom_start\n",
      "anancymaae_homolog_chrom_start\n",
      "mmulatta_homolog_chrom_start\n",
      "nscutatus_homolog_chrom_start\n",
      "pnyererei_homolog_chrom_start\n",
      "kmarmoratus_homolog_chrom_start\n",
      "cjacchus_homolog_chrom_start\n",
      "pvampyrus_homolog_chrom_start\n",
      "amexicanus_homolog_chrom_start\n",
      "mlucifugus_homolog_chrom_start\n",
      "acitrinellus_homolog_chrom_start\n",
      "munguiculatus_homolog_chrom_start\n",
      "xcouchianus_homolog_chrom_start\n",
      "mmusculus_homolog_chrom_start\n",
      "mmurinus_homolog_chrom_start\n",
      "fheteroclitus_homolog_chrom_start\n",
      "hgfemale_homolog_chrom_start\n",
      "hgmale_homolog_chrom_start\n",
      "pmbairdii_homolog_chrom_start\n",
      "elucius_homolog_chrom_start\n",
      "mmola_homolog_chrom_start\n",
      "arowi_homolog_chrom_start\n",
      "panubis_homolog_chrom_start\n",
      "mdomestica_homolog_chrom_start\n",
      "apercula_homolog_chrom_start\n",
      "pabelii_homolog_chrom_start\n",
      "ampachon_homolog_chrom_start\n",
      "cpbellii_homolog_chrom_start\n",
      "amelanoleuca_homolog_chrom_start\n",
      "pkingsleyae_homolog_chrom_start\n",
      "pmagnuspinnatus_homolog_chrom_start\n",
      "sscrofa_homolog_chrom_start\n",
      "mnemestrina_homolog_chrom_start\n",
      "oprinceps_homolog_chrom_start\n",
      "abrachyrhynchus_homolog_chrom_start\n",
      "xmaculatus_homolog_chrom_start\n",
      "oanatinus_homolog_chrom_start\n",
      "umaritimus_homolog_chrom_start\n",
      "mochrogaster_homolog_chrom_start\n",
      "ocuniculus_homolog_chrom_start\n",
      "rnorvegicus_homolog_chrom_start\n",
      "vvulpes_homolog_chrom_start\n",
      "pnattereri_homolog_chrom_start\n",
      "ecalabaricus_homolog_chrom_start\n",
      "cpugnax_homolog_chrom_start\n",
      "mcaroli_homolog_chrom_start\n",
      "scerevisiae_homolog_chrom_start\n",
      "platipinna_homolog_chrom_start\n",
      "oaries_homolog_chrom_start\n",
      "cvariegatus_homolog_chrom_start\n",
      "pmexicana_homolog_chrom_start\n",
      "saraneus_homolog_chrom_start\n",
      "mpahari_homolog_chrom_start\n",
      "bsplendens_homolog_chrom_start\n",
      "choffmanni_homolog_chrom_start\n",
      "catys_homolog_chrom_start\n",
      "apolyacanthus_homolog_chrom_start\n",
      "cpygmaea_homolog_chrom_start\n",
      "loculatus_homolog_chrom_start\n",
      "itridecemlineatus_homolog_chrom_start\n",
      "mspicilegus_homolog_chrom_start\n",
      "gaculeatus_homolog_chrom_start\n",
      "malbus_homolog_chrom_start\n",
      "csyrichta_homolog_chrom_start\n",
      "sharrisii_homolog_chrom_start\n",
      "tnigroviridis_homolog_chrom_start\n",
      "ptaltaica_homolog_chrom_start\n",
      "hcomes_homolog_chrom_start\n",
      "oniloticus_homolog_chrom_start\n",
      "csemilaevis_homolog_chrom_start\n",
      "tbelangeri_homolog_chrom_start\n",
      "xtropicalis_homolog_chrom_start\n",
      "spunctatus_homolog_chrom_start\n",
      "smaximus_homolog_chrom_start\n",
      "mgallopavo_homolog_chrom_start\n",
      "ptephrosceles_homolog_chrom_start\n",
      "ngalili_homolog_chrom_start\n",
      "csabaeus_homolog_chrom_start\n",
      "neugenii_homolog_chrom_start\n",
      "gaffinis_homolog_chrom_start\n",
      "zalbicollis_homolog_chrom_start\n",
      "bmutus_homolog_chrom_start\n",
      "sldorsalis_homolog_chrom_start\n",
      "tguttata_homolog_chrom_start\n",
      "mzebra_homolog_chrom_start\n",
      "drerio_homolog_chrom_start\n",
      "marmatus_homolog_chrom_start\n",
      "hsapiens_paralog_chrom_start\n",
      "snp_start_position\n",
      "chromosome_start\n",
      "cds_start_2076\n",
      "snp_som_start_position\n",
      "somatic_chromosome_start\n",
      "somatic_cds_start_2076\n",
      "start_exon_id\n",
      "coding_start_offset\n",
      "sequence_gene_chrom_start\n",
      "sequence_transcript_chrom_start\n",
      "sequence_transcription_start_site\n",
      "5_utr_start\n",
      "3_utr_start\n",
      "sequence_exon_chrom_start\n",
      "sequence_cdna_coding_start\n",
      "sequence_genomic_coding_start\n",
      "cds_start\n"
     ]
    }
   ],
   "source": [
    "for attribute in attributes:\n",
    "    if 'start' in attribute:\n",
    "        print(attribute)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5rc1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
