## Methods

All analysis was implemented in Python v3.8 and can be found at <https://github.com/meyer-lab/tfac-ccle>.

### Data Import & Normalization
We imported protein characterization data from the LINCS dataset, provided by Drs. Joe Gray, Laura Heiser, and James Korkola. The data came in three separate files, so they were concatenated together into one protein dataset.
	
For the gene expression dataset, we imported Oregon Health and Science University data which gave us a variety of biological markers‚Äô data. For the genetic proponent, we used the RNA sequence data given by the OHSU data.

Protein and gene expression data were normalized after importation to account for uncharacteristic outlier behavior that could influence deduced patterns from tensor factorization. (include more details)

### Gene Set Enrichment Analysis (GSEA) 
We tested for enrichment along the ARCHS4 gene list, which contains gene expression measurements from a breadth of human and mouse experiments: ARCHS4 was selected for our analysis since it is a comprehensive list of commonly tested RNA expression genes.

The ES of OHSU genes are determined by taking the maximum deviation of a running sum for each gene down the list from zero; this running sum increments by the magnitude of correlation between an observed phenotype for a given gene.

In addition to looking at ES, we want to identify genes that exhibit a low false discovery rate (FDR) in our decomposition. FDR is the ratio of false positives to the total number of true positives; by searching for genes with lower FDRs, it indicates a higher chance that the gene set is uniquely enriched for a given treatment and that it is not a false positive. For our analysis, we set FDR equal to 0.05 as the significance threshold. 

### Data Origin (LINCS Consortium)
For this project, we worked primarily with the Oregon Health and Science University‚Äôs (OHSU) protein response data as a part of the NIH Library Integrated Network-based Cellular Signatures (LINCS) program. 

The OHSU dataset was obtained from Drs. Joe Gray, Laura Heiser, and James Korkola. This was part of an NIH project to comprehensively profile how cell lines respond to a series of environmental changes. Specifically, the OHSU data consists of phenotype, protein, phosphoproteome, and RNA measurements for one cell line grown on a diverse set of microenvironments. The cells were treated with PBS, EGF, HGF, OSM, BMP2, IFNùõæ, and TGFùõΩ to assess their respective responses against each of these treatments. The effects of these signaling molecules on intracellular transcriptional- and protein-defined molecular networks will be used to study proliferation, apoptosis, differentiation state, cell binding, and motility through mathematical modeling techniques.
